<a href="https://colab.research.google.com/github/zwarshavsky/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/module1-statistics-probability-and-inference/Zhenya_Warshavsky_LS_DS_131_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 3 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.


5. You also practice some 1-sample t-tests
You try and create some kind of a visualization that communicates the results of your hypothesis tests. This can be as simple as a histogram of the p-values or the t-statistics.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [1]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2019-09-16 19:12:42--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2019-09-16 19:12:48 (610 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
#helper functions 

#returns head() of a dataframe. has an optional argument for row number. 
def h(dataframe,*args):
    return display(dataframe.head(*args),dataframe.shape)

# function for loading csv into pandas and definining the output dataframe by the file handle 
def csv(handle,file):
    globals()[handle] = pd.read_csv(file)

# function for loading csv into pandas and definining the output dataframe by the file handle and producing head(),shape for each 
def csv_h(handle,file):
    globals()[handle] = pd.read_csv(file)
    return (globals()[handle])

In [0]:
import pandas as pd
import numpy as np

In [0]:
df = pd.read_csv('house-votes-84.data',
               header=None,
               names=['party','handicapped-infants','water-project',
                        'budget','physician-fee-freeze', 'el-salvador-aid',
                        'religious-groups','anti-satellite-ban',
                        'aid-to-contras','mx-missile','immigration',
                        'synfuels', 'education', 'right-to-sue','crime','duty-free',
                        'south-africa'])

In [11]:
h(df)

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

(435, 17)

In [0]:
df = df.replace({'?':np.NaN, 'n':0, 'y':1})

In [0]:
repub = df[df["party"] == "republican"]
demo = df[df["party"] == "democrat"]

In [24]:
display(h(repub),h(demo))

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

(168, 17)

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

(267, 17)

None

None

In [80]:
#one bill where republicans appear to vote overwhelmingly in favor vs democrats
#one bill where democrats appear to vote overwhelmingly in favor vs republicans
#one bill where there appears to be an insignificant different between republican and democrat votes

column1 = "crime"
column2 = "budget"
column3 = "water-project"

display("Crime more Republican Votes","Republicans:",repub[column1].mean(),"Democrats:",demo[column1].mean(), "Budget more Democrat votes","Republicans:",repub[column2].mean(),"Democrats:",demo[column2].mean(), "Water Project marginal difference between parties","Republicans:",repub[column3].mean(),"Democrats:",demo[column3].mean())

'Crime more Republican Votes'

'Republicans:'

0.9813664596273292

'Democrats:'

0.35019455252918286

'Budget more Democrat votes'

'Republicans:'

0.13414634146341464

'Democrats:'

0.8884615384615384

'Water Project marginal difference between parties'

'Republicans:'

0.5067567567567568

'Democrats:'

0.502092050209205

In [81]:
from scipy.stats import ttest_ind

#2 sample T-test 

print("\n","Crime - Republican in Favor","\n",ttest_ind(repub[column1], demo[column1], nan_policy='omit'),"\n","Budget - Democrat in Favor","\n",ttest_ind(repub[column2], demo[column2], nan_policy='omit'),"\n","Water Project - No Difference","\n",ttest_ind(repub[column3], demo[column3], nan_policy='omit'))


 Crime - Republican in Favor 
 Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47) 
 Budget - Democrat in Favor 
 Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77) 
 Water Project - No Difference 
 Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)


###Following 5 steps of T-Test Analysis

####Crime and Budget Comparison

1) Null hypothesis - no difference in voting behvaior

2) Alternative hypothesis - major difference between voting behavior

3) Confidence Level - **95%** 

4) T-Statistic - **16.3** and **23.2** respectively

5) P-Value - **0.0000000000000000000000000000000000000000000001** and **0.00000000000000000000000000000000000000000000000000000000000000000000000000002** respectively

####Water Project Comparison

1) Null hypothesis - major difference between republican and democrat votes

2) Alternative hypothesis - no difference in voting behvaior

3) Confidence Level - **95%** 

4) T-Statistic - **.09**

5) P-Value - **0.9**



In [88]:
print("Democrats not likely to vote in favor of education bill:",demo['education'].mean())

Democrats not likely to vote in favor of education bill: 0.14457831325301204


In [96]:
from scipy.stats import ttest_1samp

print("1 Sample T-test:",ttest_1samp(demo['education'], 0, nan_policy='omit'))

1 Sample T-test: Ttest_1sampResult(statistic=6.474217555639174, pvalue=5.068051786108534e-10)


####Education 

1) Null hypothesis - There is zero democrat support for the bill

2) Alternative hypothesis - Non-zero support amongst democrats

3) Confidence Level - **95%** 

4) T-Statistic - **6.47**

5) P-Value - **0.00000000051**

In [97]:
# for converting floats

print("{:.100f}".format(float("5.068051786108534e-10")))


0.0000000005068051786108534185699238855437112616364458972384454682469367980957031250000000000000000000
